File: mobile/lib/core/models/organization.dart
/// Organization model for multi-location practice groups
///
/// Represents a parent organization that can manage multiple clinics.
class Organization {
  final String id;
  final String name;
  final String slug;
  final String? description;
  final String ownerUserId;
  final String subscriptionTier;
  final String? subscriptionExpiresAt;
  final int maxClinics;
  final int maxUsers;
  final String? logoUrl;
  final String primaryColor;
  final String? email;
  final String? phone;
  final String? website;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  // Computed fields
  final int clinicCount;
  final int userCount;

  Organization({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    required this.ownerUserId,
    required this.subscriptionTier,
    this.subscriptionExpiresAt,
    required this.maxClinics,
    required this.maxUsers,
    this.logoUrl,
    required this.primaryColor,
    this.email,
    this.phone,
    this.website,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    this.clinicCount = 0,
    this.userCount = 0,
  });

  factory Organization.fromJson(Map<String, dynamic> json) {
    return Organization(
      id: json['id'],
      name: json['name'],
      slug: json['slug'],
      description: json['description'],
      ownerUserId: json['owner_user_id'],
      subscriptionTier: json['subscription_tier'],
      subscriptionExpiresAt: json['subscription_expires_at'],
      maxClinics: json['max_clinics'],
      maxUsers: json['max_users'],
      logoUrl: json['logo_url'],
      primaryColor: json['primary_color'],
      email: json['email'],
      phone: json['phone'],
      website: json['website'],
      isActive: json['is_active'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      clinicCount: json['clinic_count'] ?? 0,
      userCount: json['user_count'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'slug': slug,
      'description': description,
      'owner_user_id': ownerUserId,
      'subscription_tier': subscriptionTier,
      'subscription_expires_at': subscriptionExpiresAt,
      'max_clinics': maxClinics,
      'max_users': maxUsers,
      'logo_url': logoUrl,
      'primary_color': primaryColor,
      'email': email,
      'phone': phone,
      'website': website,
      'is_active': isActive,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'clinic_count': clinicCount,
      'user_count': userCount,
    };
  }

  Organization copyWith({
    String? id,
    String? name,
    String? slug,
    String? description,
    String? ownerUserId,
    String? subscriptionTier,
    String? subscriptionExpiresAt,
    int? maxClinics,
    int? maxUsers,
    String? logoUrl,
    String? primaryColor,
    String? email,
    String? phone,
    String? website,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? clinicCount,
    int? userCount,
  }) {
    return Organization(
      id: id ?? this.id,
      name: name ?? this.name,
      slug: slug ?? this.slug,
      description: description ?? this.description,
      ownerUserId: ownerUserId ?? this.ownerUserId,
      subscriptionTier: subscriptionTier ?? this.subscriptionTier,
      subscriptionExpiresAt: subscriptionExpiresAt ?? this.subscriptionExpiresAt,
      maxClinics: maxClinics ?? this.maxClinics,
      maxUsers: maxUsers ?? this.maxUsers,
      logoUrl: logoUrl ?? this.logoUrl,
      primaryColor: primaryColor ?? this.primaryColor,
      email: email ?? this.email,
      phone: phone ?? this.phone,
      website: website ?? this.website,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      clinicCount: clinicCount ?? this.clinicCount,
      userCount: userCount ?? this.userCount,
    );
  }
}

File: backend/app/models/organization.py
"""
Organization model for multi-location practice management.

An Organization is a parent entity that can own multiple Clinics,
enabling hospital chains and multi-location practices.
"""

import uuid
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy import ForeignKey

from app.models.base import BaseModel, UUID

if TYPE_CHECKING:
    from app.models.clinic import Clinic
    from app.models.user import User
    from app.models.staff import StaffRole


class Organization(BaseModel):
    """
    Organization model for multi-location practice groups.

    Represents a parent entity (e.g., hospital chain, clinic network)
    that can manage multiple clinics/branches under one umbrella.

    Attributes:
        name: Organization name (e.g., "Apollo Hospitals", "HealthCare Plus")
        slug: URL-friendly unique identifier
        owner_user_id: Primary owner/admin user
        subscription_tier: Plan level (free, basic, premium, enterprise)
        max_clinics: Maximum number of clinics allowed
        max_users: Maximum staff/users allowed
        logo_url: Organization logo
        primary_color: Brand color (hex)
        is_active: Whether organization is active
        settings: JSON settings for org-wide preferences
    """

    __tablename__ = "organizations"

    # Basic Info
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    slug: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Owner
    owner_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(),
        ForeignKey("users.id"),
        nullable=False,
    )

    # Subscription & Limits
    subscription_tier: Mapped[str] = mapped_column(String(20), default="free")
    subscription_expires_at: Mapped[str | None] = mapped_column(String(50), nullable=True)
    max_clinics: Mapped[int] = mapped_column(Integer, default=1)
    max_users: Mapped[int] = mapped_column(Integer, default=5)

    # Branding
    logo_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    primary_color: Mapped[str] = mapped_column(String(7), default="#1976D2")  # Hex color

    # Contact
    email: Mapped[str | None] = mapped_column(String(255), nullable=True)
    phone: Mapped[str | None] = mapped_column(String(15), nullable=True)
    website: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # Status
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)

    # Relationships
    clinics: Mapped[list["Clinic"]] = relationship(
        "Clinic",
        back_populates="organization",
        foreign_keys="Clinic.organization_id",
    )
    staff_roles: Mapped[list["StaffRole"]] = relationship(
        "StaffRole",
        back_populates="organization",
        cascade="all, delete-orphan",
    )

    def __repr__(self) -> str:
        return f"<Organization {self.name}>"
